runner/scheduler.py

```diff
@@ -35,86 +37,83 @@ def are_compatible(a: TestExample, b: TestExample, incompatibilities: list[set[t
         statement_times,
         datetime.now(),
     )
 
 
-def extract_log_kwargs(action_log: list[TestAction]):
-    kwargs = dict()
-    for k in ["questions", "question_responses", "messages_timestamps", "task_log"]:
-        kwargs[k] = []
-    for action in action_log:
-        if not isinstance(action, SendMessageAction):
-            continue
-        kwargs["task_log"].append(f"Test ({action.sent_ts}): " + action.message)
-        kwargs["task_log"].append(f"Agent ({action.reply_ts}): " + action.reply)
-        if not action.is_filling:
-            kwargs["messages_timestamps"].append(action.sent_ts)
-        if action.is_question:
-            kwargs["questions"].append(action.message)
-            kwargs["question_responses"].append(action.reply)
-    return kwargs
-
 
 @dataclass
 class TestRunner:
     config: RunConfig
     agent: ChatSession
     tests: list[TestExample]
-    results: list[TestResult] = field(default_factory=list)
+    finished_results: list[TestResult] = field(default_factory=list)
+    in_progress_results: dict[str, TestResult] = field(default_factory=dict)
     traveller: Optional[time_machine.travel] = None
     wait_list: dict[str, dict[str, int | datetime]] = field(default_factory=dict)
     current_token_count: int = 0
     avg_tokens_per_second: Optional[float] = None
-    action_logs: dict[str, list[TestAction]] = field(default_factory=dict)
     test_managing_costs_usd: float = 0
     benchmark_duration_seconds: float = 0
     reference_duration_timestamp: Optional[datetime] = None
     skip_evaluations: bool = False
-    result_callbacks: List[Tuple[Callable, TestExample, TestResult]] = field(default_factory=list)
-    master_log: List[str] = field(default_factory=list)
+    result_callbacks: List[Tuple[Callable, TestExample]] = field(default_factory=list)
+    master_log: MasterLog = field(default_factory=lambda: MasterLog())
     progress_dialog: ProgressDialog = None
 
     @property
-    def saving_path(self):
+    def runstats_path(self):
         return make_runstats_path(self.config.run_name, self.agent.name)
 
+    @property
+    def master_log_path(self):
+        return make_master_log_path(self.config.run_name, self.agent.name)
+
     def save(self):
         self.update_duration()
-        self.saving_path.parent.mkdir(parents=True, exist_ok=True)
+        self.runstats_path.parent.mkdir(parents=True, exist_ok=True)
         stats = dict(
             total_tokens=self.current_token_count,
             agent_costs_usd=self.agent.costs_usd,
             managing_costs_usd=self.test_managing_costs_usd,
             tokens_per_second=self.avg_tokens_per_second,
             duration=self.benchmark_duration_seconds,
         )
-        with open(self.saving_path, "w") as fd:
+
+        with open(self.runstats_path, "w") as fd:
             json.dump(stats, fd)
 
+        with open(self.master_log_path, "w") as fd:
+            json.dump(self.master_log.to_json(), fd, indent=2)
```

Comment:
I've been thinking about the saving frequency we should keep now that we have the master log. Right now it's being saved every time a test finishes, which makes it not so different from what we had before. Ideally, we would save after every agent response, but that could get us in trouble when logs grow, so I thought about keeping the master log as a `jsonl` file (every line is a JSON object). If we did that, we could save a new event by just appending to the file, and the runstats file should be small enough to save it every so often. What do you think?